### PR TITLE
Type inference bug fix

### LINKED ISSUE #1437: Type promotion does not match with eager mode
### 🐛 Describe the bug

```python
import torch

torch._C._jit_set_nvfuser_enabled(True)
torch._C._jit_set_texpr_fuser_enabled(False)
torch._C._jit_set_profiling_executor(True)
torch._C._jit_set_profiling_mode(True)

x = torch.tensor(0, device='cuda:0', dtype=torch.float16)
y = torch.tensor(0, dtype=torch.float32)
z = torch.tensor([0], device='cuda:0', dtype=torch.float16)
q = torch.tensor(0, device='cuda:0', dtype=torch.float)

print("========== case 1 ==========")
o = x * y + z
print("Eager:", o.dtype)

@torch.jit.script
def f(x, y, z):
    return x * y + z

o = f(x, y, z)
print("JIT 1:", o.dtype)
o = f(x, y, z)
print("JIT 2:", o.dtype)
o = f(x, y, z)
print("JIT 3:", o.dtype)
```

I am seeing
```
========== case 1 ==========
Eager: torch.float16
JIT 1: torch.float16
JIT 2: torch.float32
JIT 3: torch.float32
```

### PR DESCRIPTION
**Pull request by zasdfgbnm**
It is important that the type inference code compute dim correctly because type promotion logic uses dim information. Currently, our type inference code just assigning `c10::nullopt` for dim, which causes bugs not detected by unit tests.

The unit tests that are supposed to test these cases were skipped because of the poor numerics of half, and bfloat16. In this PR, I reenable these tests, but only for dtypes.

I am pretty sure this PR does not fix all bugs, I will submit follow-up PRs for the remaining bugs.

Fixes https://github.com/csarofeen/pytorch/issues/1437

### CODE REVIEW

**Review by rdspring1 — APPROVED** [review]
LGTM

**Comment by jjsjann123** [comment]
> It is important that the type inference code compute dim correctly because type promotion logic uses dim information. Currently, our type inference code just assigning c10::nullopt for dim, which causes bugs not detected by unit tests.

Dumb question, did changes in type inference actually fixes anything??

> our type inference code just assigning c10::nullopt for dim,

This is done explicitly by design, we used to have shape propagation in this function, but that was pulled out since we now uses this for all shape propagation: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/manager.cpp#L185

**Comment by zasdfgbnm** [comment]
> Dumb question, did changes in type inference actually fixes anything??

It does. The fixed bug is https://github.com/csarofeen/pytorch/issues/1437. The root cause for this bug is the following:

In `at::native::result_type`
```c++
ScalarType result_type(const ResultTypeState& in_state) {
  return combine_categories(in_state.dimResult, combine_categories(in_state.zeroResult, in_state.wrappedResult));
}
```
it combines the result from `dimResult`, `zeroResult`, and `wrappedResult`, and in the type inference code, we have
```C++
at::native::ResultTypeState updateResultTypeState(
    OperandType tensor,
    const at::native::ResultTypeState& in_state) {
  at::native::ResultTypeState new_state = in_state;
  c10::ScalarType current = tensor.scalar_type;

  if (tensor.dim > 0) {
    new_state.dimResult =
        promoteTypesSkipUndefined(in_state.dimResult, current);
  } else {
    new_state.zeroResult =
        promoteTypesSkipUndefined(in_state.zeroResult, current);
  }
  return new_state;
}
```
without `tensor.dim` correctly set,  `updateResultTypeState` will incorrectly set `dimResult` instead of `zeroResult`, which causes wrong results.

> This is done explicitly by design, we used to have shape propagation in this function, but that was pulled out since we now uses this for all shape propagation: https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/manager.cpp#L185

If this is the case, I think we need to pull the dim info from shape propagation and use that info to set the `dim` correctly. But how do I access these information? I remember trying to print something like `node->output()->cast<TensorType>()->dim()` inside `binary_type` and get `nullopt`.

**Comment by jjsjann123** [comment]
Thanks for pointing out the context, forgot that we have scalar tensor have special type promotion logic. :face_exhaling: 

So the reason that shape information are wiped out is because that type_inference explicitly overwrites shape information. If you look at the pattern we have here, you'll see lots of `TensorType::create(*op->scalarType(), *op->device(), c10::nullopt, c10::nullopt);`, those should all be updated to forward the existing shape/stride that has already been on the tensor types.

One more thing to add is to clean up the shape propagation pass, I think the one after type inference can be removed as well.
https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/manager.cpp#L185

I do like the updates despite this `computeDimForBinaryOp`. :stuck_out_tongue_closed_eyes: 

Sorry this starts to look a lot more like code cleaning that's due on our part. How do you want to proceed this? Would be more than welcomed if you want to clean type_inference part up. Otherwise, I can push a refactor to your PR directly (ETA would be ~ end of Thursday / Friday morning).

**Comment by zasdfgbnm** [comment]
I will do the cleanup in this PR directly. And I will ping you if I need help :)

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
marking request changes to avoid accidental merge.

**Comment by csarofeen** [comment]
Requesting re-review to reflect requested changes.

**Comment by zasdfgbnm** [comment]
This should be ready

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 1]
```diff
@@ -29,6 +29,16 @@ bool hasTypeAndDevice(const TensorTypePtr& op) {
       op->scalarType().has_value();
 }
 
+void copyScalarTypeAndDeviceToOutput(
+    TensorTypePtr from,
+    Node* node,
+    size_t index = 0) {
+  auto out = node->output(index)->type()->cast<TensorType>();
+  TORCH_INTERNAL_ASSERT(out != nullptr);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: nitpick: would be nice to have a message attached.

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 2]
```diff
@@ -528,33 +518,39 @@ class NaiveTypePropagator {
   }
 
  protected:
-  TensorTypePtr unary_type(Node* node) {
+  void unary_type(Node* node) {
     auto op = getInputTensorType(node, 0, false);
-    return TensorType::create(
-        *op->scalarType(), *op->device(), c10::nullopt, c10::nullopt);
+    auto out = node->output()->type()->cast<TensorType>();
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: nitpick: this could be changed to `copyScalarTypeAndDeviceToOutput(op, out)`.

File: torch/csrc/jit/codegen/cuda/type_inference.cpp [hunk 3]
```diff
@@ -528,33 +518,39 @@ class NaiveTypePropagator {
...
+      Node* node,
       const TensorTypePtr& op,
       const std::vector<int64_t>& dims,
       bool keepdim) {
     TORCH_CHECK(
         hasTypeAndDevice(op),
         "Type and device propagation has failed, or was not provided enough information.");
-    return TensorType::create(
-        *op->scalarType(), *op->device(), c10::nullopt, c10::nullopt);
+    auto out = node->output()->type()->cast<TensorType>();
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/type_inference.cpp]: ditto.

**Review by jjsjann123 — APPROVED** [review]
Thanks a lot for cleaning up for us :heart: 
LGTM
